docs/kubeconfig.md
## API server access

Access to the Kubernetes API server is needed:
* To create the virtual node object, watch pods, and list various Kubernetes resources in Kip. 
* To allow Kip to create and manage its CRDs.
* By the network agent, running on cells, to be able to implement service proxying and manage network policies.

By default, Kubernetes service accounts are used for authentication and authorization. The Kip pod service account is used for the first two use cases, and another service account is used by the cell network agent. If you use the [example deployment manifests and configs](deploy/) provided, the necessary service accounts and cluster roles will be created and configured at setup time.

As another option, both Kip and the network agent can use a kubeconfig file (either the same, or separate ones).

You can use a ConfigMap to attach a kubeconfig file into the container:

    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: kip-config-cfd95dk474
      namespace: kube-system
    data:
      kubeconfig: |
        apiVersion: v1
        clusters:
        - cluster:
            certificate-authority-data: <base64-encoded CA certificate>
            server: https://10.20.30.40:6443 # API server.
          name: kubernetes
        contexts:
        - context:
            cluster: kubernetes
            user: kubernetes-admin
          name: kubernetes-admin@kubernetes
        current-context: kubernetes-admin@kubernetes
        kind: Config
        preferences: {}
        users:
        - name: kubernetes-admin
          user:
            client-key-data: <base64-encoded client key>
            client-certificate-data: <base64-encoded client certificate>

Create a volume of the file in the ConfigMap:

    volumes:
    - configMap:
        defaultMode: 420
        items:
        - key: kubeconfig
          mode: 384
          path: kubeconfig
        name: kip-config-cfd95dk474
      name: kip-config

Attach the volume to the container:

    containers:
    - name: kip
      [...]
      volumeMounts:
      - mountPath: /etc/kip
        name: kip-config

To configure access via a kubeconfig file, make sure the kubeconfig file is available and set the `KUBECONFIG` environment variable for the Kip container:

    containers:
    - name: kip
      [...]
      env:
      - name: KUBECONFIG
        value: /etc/kip/kubeconfig

For the network agent, use the command line parameter `--network-agent-kubeconfig`, for example:

    containers:
    - name: kip
      command:
      - /kip
      - --provider
      - kip
      - --provider-config
      - /etc/kip/provider.yaml
      - --network-agent-kubeconfig
      - /etc/kip/kubeconfig
      env:
      - name: KUBECONFIG
        value: /etc/kip/kubeconfig


docs/provider-config.md
## Provider Configuration

Kip is configured using a yaml file.  The easiest way to get that file into kip’s virtual-kubelet pod is to use a ConfigMap and then supply the location of the file to kip via the `--provider-config` flag.  A sample config file with documentation comments for each section is shown below:

```yaml
---
# Like all the resources in k8s, this configuration file is versioned.
apiVersion: v1

# configures the cloud provider for kip
cloud:

  # Only one cloud provider can be enabled in kip. Currently
  # AWS and GCE are supported. We are working on support for
  # Azure

  aws:

    # The AWS region where kip will create instances. Example:
    # `us-east-1`. The environment variable `AWS_REGION` can also be
    # used instead.
    region: us-east-1

    # The AWS access key ID Kip will use for interacting with the
    # AWS API. The environment variable `AWS_ACCESS_KEY_ID` can also
    # be used instead.
    accessKeyID: FILL_IN

    # The AWS secret access key. The environment variable
    # `AWS_SECRET_ACCESS_KEY` can also be used instead.
    secretAccessKey: FILL_IN

    # This is the VPC ID in which instances will be created.
    # "default" will selecct the default VPC. If empty, kip will
    # attempt to detect if it is running on an instance inside a VPC
    # (via AWS metadata) and will use its current VPC.
    # vpcID: ''

    # the AWS subnet ID of the subnet kip will launch pods into.  if
    # blank, kip will detect its current subnet (via AWS metadata)
    # and will launch pods into that subnet.
    # subnetID: ''

    # Specify a custom endpoint URL for AWS EC2 service/API requests
    # this corresponds to the --endpoint-url option in the AWS CLI
    # endpointURL: ''

    # Disable verifying the API server's certificate chain and host
    # name. In this mode, TLS is susceptible to man-in-the-middle
    # attacks. Disabling security checks is dangerous and should be
    # avoided. This should only be used for testing.
    # insecureTLSSkipVerify: false

  # gce:

    # The GCE project where kip will be running. Can be auto-detected
    # using instance metadata
    # projectID: "my-project"

    # Kip can use the service account attached to the GCE instance it runs on.
    # If the instance doesn't have the https://www.googleapis.com/auth/compute
    # scope attached then credentials for kip to launch instances can be
    # manually supplied via a clientEmail and service account private key
    # credentials:
      # clientEmail: my-account@my-project.iam.gserviceaccount.com
      # privateKey: "-----BEGIN PRIVATE KEY-----\n[base64-encoded private key]-----END PRIVATE KEY-----\n"

    # The zone where kip will run and launch instance into. Can be
    # auto-detected using instance metadata.
    # zone: us-central1-c

    # The name of the virtual cloud network where kip will run. Can be
    # auto-detected using instance metadata.
    # vpcName: "default"

    # The name of the subnet where kip will run and launch instance into.
    # Can be auto-detected using instance metadata.
    # subnetName: "default"

# the etcd section controls how kip stores its state, either using
# an external etcd cluster or using an embedded etcd database.
etcd:
  # Settings for running kip from an embedded etcd server.
  internal:

    # The dataDir is the directory that will be used by etcd for
    # storage. The kip user must have write access to the directory.
    # Defaults to `/opt/kip/data`. The data directory can also be
    # specified in the `configFile` but should not be specified in both
    # locations.
    dataDir: "/opt/kip/data"

  # Path to an etcd configuration file that will be used to further
  # customize the behavior of etcd. All etcd command-line flags are
  # supported. For more information, see the documentation available
  # on the etcd website.
  # configFile: /opt/kip/etc/etcd.yml

# the kubelet section describes the specs of the virtual-kubelet node
# that will be created by kip. This is the advertised size of the
# kubernetes node, not the size of of the physical instance the
# virtual-kubelet is running on. Values are resource quantities,
# similar to those found in a container's Resources configuration.

kubelet:
  cpu: "64"
  memory: "512Gi"
  pods: "400"

# the cells section configures parameters that affect kip cells
cells:
  # nametag is a name that will be added onto cloud tags to help
  # identify cloud resources that belong to this controller. Because
  # of restrictions across clouds, the nametag must be a valid dns
  # label (the name must start with a lower case letter followed by
  # lowercase letters numbers or dashes) and must have a length of 34
  # characters or less.
  nametag: kip

  # defaultInstanceType specifies the the default cloud instance type
  # Kip will use when creating pods, if the pod does not set an
  # instance type. Examples t3.micro, c4.8xlarge, p2.xlarge
  defaultInstanceType: t3.nano

  # defaultVolumeSize specifies the default size for the root volume
  # of cells in bytes.  To set the root volume to 8 GiB, specify the
  # value as "8Gi".  Must be 1Gi or larger, defaults to 5Gi.
  defaultVolumeSize: "5Gi"

  # defaultIAMPermissions specifies the a cloud-specific IAM permission that
  # will be attached to cell instances. On AWS, this is an IAM instance
  # profile, and can be overridden by the pod.elotl.co/instance-profile pod
  # annotation.
  #defaultIAMPermissions: "arn:aws:iam::11123456789:instance-profile/kip-cell"

  # bootImageSpec is a dictionary of cloud-specific image properties for
  # specifying the boot image to use for cells.
  # Valid fields on AWS are:
  #   - owners, which is a space separated list of AWS account IDs, "self", or
  #     an AWS owner alias such as "amazon" or "aws-marketplace".
  #   - executableUsers, which is a space separated list of users with explicit
  #     launch permissions. You can specify an AWS account ID, "self" or "all".
  #   - filters, which lets you filter results. Example:
  #     "name=elotl-kip-* is-public=true". This is equivalent to using:
  #     $ aws ec2 describe-images \
  #     --filters Name=name,Values=elotl-kip-*,Name=is-public,Values=true
  #   - imageIDs, one or more space separated image IDs. Example:
  #     "ami-07cfdf0f7c08293fd"
  # If there are multiple matching images, the latest one will be used.
  # See
  # https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-images.html
  # for more details.
  # The below settings are the default if no bootImageSpec is specified.
  bootImageSpec:
    owners: "689494258501"
    filters: "name=elotl-kip-*"

  # cloudInitFile specifies a path to a cloudInitFile that will be
  # used to provision all cells that Kip boots. Kip will detect
  # modifications to this file. Cells started afte a modification are
  # made will get the updated cloudInit file.
  #
  # cloudInitFile: /etc/kip/cloudinit.yml

  # standbyCells is used to speicfy pools of standby cells kip will
  # keep so pods created can be dispatched to cells quickly.
  # standbyCells should be a list of `standbyCell` values that contain
  # three parameters
  #  * instanceType: the name of the cloud instance type
  #  * spot: whether the instance should be a spot instance
  #  * count: the number of standby instances of this types
  #
  # standbyCells:
  #   - instanceType: "t2.micro"
  #     count: 2
  #     spot: false

  # extraSecurityGroups contains the IDs of additional security groups
  # that will be attached to kip cells.  In AWS those groups looks
  # like: sg-av3sp192jur
  #
  # extraSecurityGroups:
  #   - sg-246810

  # By default, cells will be assigned a publicIP address if the
  # subnet is configured to allow access to the public internet
  # without NAT.  Set privateIPOnly to true to force all cells
  # to only have a private IP address.

  # privateIPOnly: false

  # itzo configures the version of the itzo agent to use and where
  # itzo will be downloaded from.  You should only customize this if
  # you have built your own itzo agent or you would like to pin your
  # itzo agent version to a particular version.

  # itzo:
  #   version: 532
  #   url: "http://itzo-kip-download.s3.amazonaws.com"

  # statusInterval controls how often (in seconds) kip will query cells
  # for their status.

  # statusInterval: 5

  # the healthCheck section is used to configure how cells are
  # health checked.  Kip supports two methods of determining cell health:
  # 1) Status health checks that ensure cells are responding to
  #    status queries from kip.
  # 2) Cloud API healthchecks supplement status checks by also
  #    querying the cloud API to determine cell health based on the
  #    status of the cell's backing compute instance.

  # status is the default healthcheck mode and is the more robust of
  # the two methods. cloudAPI health checks are useful for
  # configurations where where kip running outside the cloud network
  # and connected to cells through the internet. Kip will query the
  # cloud provider's API to determine if a cell is running. cloudAPI
  # healthchecks provide less confidence the cells are functioning
  # correctly but allows cells to survive temporary network
  # connectivity issues between the kip controller and the cells.
  # Only one of status or cloudAPI healthcheck can be specified in the
  # healthcheck section. When cloudAPI health checks are used, status
  # queries are still used to check cell status.

  # healthcheck:
  #   status:
  #     healthyTimeout: 90
  #   cloudAPI:
  #     healthyTimeout: 180
  #     # interval between checks must be >= 10, default is 60
  #     interval: 60

  # CellConfig can be used to send a map of key/value pairs to cells. It is
  # used to configure the image cache right now.
  # cellConfig:
  #   imageCacheEndpoint: 10.20.0.2:/data
  #   imageCacheMountDir: /var/cache/images
  #   imageCacheMountOpts: -o ro
```


pkg/server/config.go
/*
Copyright 2020 Elotl Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package server

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"

	"github.com/elotl/kip/pkg/api"
	"github.com/elotl/kip/pkg/api/validation"
	"github.com/elotl/kip/pkg/server/cloud"
	"github.com/elotl/kip/pkg/server/cloud/aws"
	"github.com/elotl/kip/pkg/server/cloud/azure"
	"github.com/elotl/kip/pkg/server/cloud/gce"
	"github.com/elotl/kip/pkg/server/nodemanager"
	"github.com/elotl/kip/pkg/util"
	vutil "github.com/elotl/kip/pkg/util/validation"
	"github.com/elotl/kip/pkg/util/validation/field"
	"github.com/elotl/kip/pkg/util/yaml"
	v1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/resource"
	"k8s.io/klog"
)

const (
	blankTemplateValue = "FILL_IN"
)

var (
	defaultStatusInterval              = 5
	defaultCloudAPIHealthCheckInterval = 60
	defaultCloudAPIHealthCheckTimeout  = 180
	defaultStatusHealthCheckTimeout    = 90

	defaultCPUCapacity    = resource.MustParse("20")
	defaultMemoryCapacity = resource.MustParse("100Gi")
	defaultPodCapacity    = resource.MustParse("20")
)

// ServerConfigFile stores the parsed json from provider.yaml
type ServerConfigFile struct {
	api.TypeMeta `json:",inline"`
	Cloud        MultiCloudConfig `json:"cloud"`
	Etcd         EtcdConfig       `json:"etcd"`
	Cells        CellsConfig      `json:"cells"`
	Testing      TestingConfig    `json:"testing"`
	Kubelet      KubeletConfig    `json:"kubelet"`
}

// Kubelet stores kubelet-specific configuration such as capacity and labels.
type KubeletConfig struct {
	// Deprecated: CPU, Memory and Pods are copied into Capacity, and are only
	// present for backward compatibility.
	CPU      *resource.Quantity `json:"cpu"`
	Memory   *resource.Quantity `json:"memory"`
	Pods     *resource.Quantity `json:"pods"`
	Capacity v1.ResourceList    `json:"capacity"`
	Labels   map[string]string  `json:"labels"`
}

type MultiCloudConfig struct {
	AWS   *AWSConfig   `json:"aws,omitempty"`
	GCE   *GCEConfig   `json:"gce,omitempty"`
	Azure *AzureConfig `json:"azure,omitempty"`
}

type AWSConfig struct {
	Region                string `json:"region"`
	AccessKeyID           string `json:"accessKeyID"`
	SecretAccessKey       string `json:"secretAccessKey"`
	VPCID                 string `json:"vpcID,omitempty"`
	SubnetID              string `json:"subnetID,omitempty"`
	EcsClusterName        string `json:"ecsClusterName"`
	EndpointURL           string `json:"endpointURL"`
	InsecureTLSSkipVerify bool   `json:"insecureTLSSkipVerify"`
}

// See https://github.com/Azure/azure-sdk-for-go/blob/master/README.md
// for more info on SDK login credentials.  We might want to support
// CertificatePath and CertificatePassword.
type AzureConfig struct {
	SubscriptionID string `json:"subscriptionID"`
	Location       string `json:"location"`
	VNetName       string `json:"virtualNetworkName"`
	SubnetName     string `json:"subnetName"`
	TenantID       string `json:"tenantID"`
	ClientID       string `json:"clientID"`
	ClientSecret   string `json:"clientSecret"`
}

type GCEConfig struct {
	ProjectID       string          `json:"projectID"`
	CredentialsFile string          `json:"credentialsFile,omitempty"`
	Credentials     *GCECredentials `json:"credentials,omitempty"`
	Zone            string          `json:"zone,omitempty"`
	VPCName         string          `json:"vpcName,omitempty"`
	SubnetName      string          `json:"subnetName,omitempty"`
}

type GCECredentials struct {
	ClientEmail string `json:"clientEmail"`
	PrivateKey  string `json:"privateKey"`
}

type EtcdConfig struct {
	Client   EtcdClientConfig   `json:"client"`
	Internal InternalEtcdConfig `json:"internal"`
}

type EtcdClientConfig struct {
	Endpoints []string `json:"endpoints"`
	CertFile  string   `json:"certFile"`
	KeyFile   string   `json:"keyFile"`
	CAFile    string   `json:"caFile"`
}

type InternalEtcdConfig struct {
	DataDir    string `json:"dataDir"`
	ConfigFile string `json:"configFile"`
}

type CellsConfig struct {
	BootImageSpec         cloud.BootImageSpec           `json:"bootImageSpec"`
	DefaultInstanceType   string                        `json:"defaultInstanceType"`
	DefaultVolumeSize     string                        `json:"defaultVolumeSize"`
	StandbyCells          []nodemanager.StandbyNodeSpec `json:"standbyCells"`
	CloudInitFile         string                        `json:"cloudInitFile"`
	Itzo                  ItzoConfig                    `json:"itzo"`
	ExtraCIDRs            []string                      `json:"extraCIDRs"`
	ExtraSecurityGroups   []string                      `json:"extraSecurityGroups"`
	Nametag               string                        `json:"nametag"`
	StatusInterval        int                           `json:"statusInterval"`
	HealthCheck           HealthCheckConfig             `json:"healthcheck"`
	PrivateIPOnly         *bool                         `json:"privateIPOnly"`
	CellConfig            map[string]string             `json:"cellConfig"`
	DefaultIAMPermissions string                        `json:"defaultIAMPermissions"`
}

type HealthCheckConfig struct {
	Status   *StatusHealthCheck   `json:"status"`
	CloudAPI *CloudAPIHealthCheck `json:"cloudAPI"`
}

type StatusHealthCheck struct {
	HealthyTimeout int `json:"healthyTimeout"`
}

type CloudAPIHealthCheck struct {
	Interval       int `json:"interval"`
	HealthyTimeout int `json:"healthyTimeout"`
}

type ItzoConfig struct {
	Version string `json:"version"`
	URL     string `json:"url"`
}

type TestingConfig struct {
	ControllerID string `json:"controllerID"`
}

func serverConfigFileWithDefaults() *ServerConfigFile {
	sc := ServerConfigFile{
		TypeMeta: api.TypeMeta{
			Kind:       "serverConfig",
			APIVersion: "v1",
		},
		Etcd: EtcdConfig{
			Internal: InternalEtcdConfig{
				DataDir: "/opt/milpa/data",
			},
		},
		Cells: CellsConfig{
			BootImageSpec:     cloud.BootImageSpec{},
			StandbyCells:      []nodemanager.StandbyNodeSpec{},
			DefaultVolumeSize: "5Gi",
			StatusInterval:    defaultStatusInterval,
		},
		Kubelet: KubeletConfig{
			Capacity: v1.ResourceList{
				"cpu":    defaultCPUCapacity,
				"memory": defaultMemoryCapacity,
				"pods":   defaultPodCapacity,
			},
			Labels: map[string]string{},
		},
	}
	return &sc
}

func setEnvIfNotSet(envVar, value string) error {
	if os.Getenv(envVar) == "" && value != "" {
		err := os.Setenv(envVar, value)
		if err != nil {
			msg := fmt.Sprintf("Error setting %s env var", envVar)
			return util.WrapError(err, msg)
		}
	}
	return nil
}

func setupAWSRegion(configRegion string) error {
	winningRegionVal := configRegion
	if os.Getenv("AWS_REGION") != "" {
		winningRegionVal = os.Getenv("AWS_REGION")
	}
	if os.Getenv("AWS_DEFAULT_REGION") != "" {
		winningRegionVal = os.Getenv("AWS_DEFAULT_REGION")
	}
	klog.V(2).Infof("using AWS region %q", winningRegionVal)
	return os.Setenv("AWS_REGION", winningRegionVal)
}

func setupAwsEnvVars(c *AWSConfig) error {
	if err := setEnvIfNotSet("AWS_ACCESS_KEY_ID", c.AccessKeyID); err != nil {
		return err
	}
	if err := setEnvIfNotSet("AWS_SECRET_ACCESS_KEY", c.SecretAccessKey); err != nil {
		return err
	}
	region := c.Region
	if region == "" {
		region = aws.AutoDetectRegion()
	}
	if region != "" {
		if err := setupAWSRegion(region); err != nil {
			return err
		}
	}
	klog.V(2).Infof("Validating connection to AWS")
	if err := aws.CheckConnection(c.EndpointURL, c.InsecureTLSSkipVerify); err != nil {
		return util.WrapError(err, "Error validationg connection to AWS")
	}
	klog.V(2).Infof("Validated access to AWS")
	return nil
}

func setupAzureEnvVars(c *AzureConfig) error {
	if err := setEnvIfNotSet("AZURE_TENANT_ID", c.TenantID); err != nil {
		return err
	}
	if err := setEnvIfNotSet("AZURE_CLIENT_ID", c.ClientID); err != nil {
		return err
	}
	if err := setEnvIfNotSet("AZURE_CLIENT_SECRET", c.ClientSecret); err != nil {
		return err
	}
	if err := os.Setenv("SUBSCRIPTION_ID", c.SubscriptionID); err != nil {
		return err
	}
	klog.V(2).Infof("Validating connection to Azure")
	if err := azure.CheckConnection(c.SubscriptionID); err != nil {
		return util.WrapError(err, "Error validationg connection to Azure")
	}
	klog.V(2).Infof("Validated access to Azure")
	return nil
}

func configureCloudProvider(cf *ServerConfigFile, controllerID, nametag string) (cloud.CloudClient, error) {
	// see which cloud is non-null, take first
	mockCloudAPI := os.Getenv("MOCK_CLOUD_API")
	if val, err := strconv.ParseBool(mockCloudAPI); err == nil && val {
		klog.Warningf("Running with a mocked cloud API client. This kip installation can not be changed to run a real k8s cluster")
		return cloud.NewMockClient(), nil
	}
	cc := cf.Cloud
	numClouds := 0
	if cc.AWS != nil {
		numClouds++
	}
	if cc.Azure != nil {
		numClouds++
	}
	if cc.GCE != nil {
		numClouds++
	}
	if numClouds > 1 {
		return nil, fmt.Errorf("Multiple clouds configured in cloud section of provider.yaml")
	}
	privateIPOnly := false
	if cf.Cells.PrivateIPOnly != nil && *cf.Cells.PrivateIPOnly {
		privateIPOnly = true
	}
	if cc.AWS != nil {
		errs := validateAWSConfig(cc.AWS)
		if len(errs) > 0 {
			err := fmt.Errorf("Invalid AWS Cloud Config: %v", errs.ToAggregate())
			return nil, err
		}

		err := setupAwsEnvVars(cc.AWS)
		if err != nil {
			return nil, util.WrapError(err, "Could not configure AWS cloud client authorization")
		}

		// Gross: if vpc is "default", the NewEC2Client will
		// attempt to figure out the VPCID and the actual ID
		// will be available from there

		client, err := aws.NewEC2Client(aws.EC2ClientConfig{
			ControllerID:          controllerID,
			Nametag:               nametag,
			VPCID:                 cc.AWS.VPCID,
			SubnetID:              cc.AWS.SubnetID,
			ECSClusterName:        cc.AWS.EcsClusterName,
			PrivateIPOnly:         privateIPOnly,
			EndpointURL:           cc.AWS.EndpointURL,
			InsecureTLSSkipVerify: cc.AWS.InsecureTLSSkipVerify,
		})

		if err != nil {
			return nil, util.WrapError(err, "Error creating AWS cloud client")
		}
		return client, nil
	} else if cc.Azure != nil {
		errs := validateAzureConfig(cc.Azure)
		if len(errs) > 0 {
			err := fmt.Errorf("Invalid Azure Cloud Config: %v", errs.ToAggregate())
			return nil, err
		}
		err := setupAzureEnvVars(cc.Azure)
		if err != nil {
			return nil, util.WrapError(err, "Could not configure Azure cloud client")
		}
		client, err := azure.NewAzureClient(
			controllerID,
			nametag,
			cc.Azure.SubscriptionID,
			cc.Azure.Location,
			cc.Azure.VNetName,
			cc.Azure.SubnetName,
		)
		if err != nil {
			return nil, util.WrapError(err, "Error creating Azure cloud client")
		}
		return client, nil
	} else if cc.GCE != nil {
		errs := validateGCEConfig(cc.GCE)
		if len(errs) > 0 {
			err := fmt.Errorf("Invalid GCE Cloud Config: %v", errs.ToAggregate())
			return nil, err
		}
		options := make([]gce.ClientOption, 0, 4)
		options = append(options, gce.WithZone(cc.GCE.Zone))
		options = append(options, gce.WithVPCName(cc.GCE.VPCName))
		options = append(options, gce.WithSubnetName(cc.GCE.SubnetName))
		options = append(options, gce.WithPrivateIPOnly(privateIPOnly))
		if cc.GCE.CredentialsFile != "" {
			options = append(options, gce.WithCredentialsFile(cc.GCE.CredentialsFile))
		}
		if cc.GCE.Credentials != nil {
			options = append(options, gce.WithCredentials(cc.GCE.Credentials.ClientEmail, cc.GCE.Credentials.PrivateKey))
		}
		client, err := gce.NewGCEClient(controllerID, nametag, cc.GCE.ProjectID, options...)
		if err != nil {
			return nil, util.WrapError(err, "Error creating GCE cloud client")
		}
		return client, nil
	} else {
		return nil, fmt.Errorf("You must specify a cloud configuration in provider.yaml")
	}
}

func ParseConfig(path string) (*ServerConfigFile, error) {
	var err error

	// unmarshal into ServerConfigFile
	if _, err = os.Stat(path); os.IsNotExist(err) {
		return nil, fmt.Errorf("Missing config file %s", path)
	}
	configData, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, util.WrapError(err, "Could not read provider.yaml")
	}
	configFile := serverConfigFileWithDefaults()
	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(configData), bufferSize)
	err = decoder.Decode(&configFile)
	if err != nil {
		return nil, util.WrapError(err, "Error parsing provider.yaml")
	}

	setConfigDefaults(configFile)

	updateCapacityFromDeprecatedFields(configFile)

	return configFile, nil
}

func updateCapacityFromDeprecatedFields(config *ServerConfigFile) {
	if config.Kubelet.CPU != nil {
		config.Kubelet.Capacity[v1.ResourceCPU] = *config.Kubelet.CPU
	}
	if config.Kubelet.Memory != nil {
		config.Kubelet.Capacity[v1.ResourceMemory] = *config.Kubelet.Memory
	}
	if config.Kubelet.Pods != nil {
		config.Kubelet.Capacity[v1.ResourcePods] = *config.Kubelet.Pods
	}
}

// Sets default values for parameters that can only be set once the
// ServerConfigFile has been parsed
func setConfigDefaults(config *ServerConfigFile) {
	if config.Cells.HealthCheck.Status == nil && config.Cells.HealthCheck.CloudAPI == nil {
		config.Cells.HealthCheck = HealthCheckConfig{
			Status: &StatusHealthCheck{
				HealthyTimeout: defaultStatusHealthCheckTimeout,
			},
		}
	}
	if config.Cells.HealthCheck.Status != nil {
		if config.Cells.HealthCheck.Status.HealthyTimeout == 0 {
			config.Cells.HealthCheck.Status.HealthyTimeout = defaultStatusHealthCheckTimeout
		}
	} else if config.Cells.HealthCheck.CloudAPI != nil {
		if config.Cells.HealthCheck.CloudAPI.HealthyTimeout == 0 {
			config.Cells.HealthCheck.CloudAPI.HealthyTimeout = defaultCloudAPIHealthCheckTimeout
		}
		if config.Cells.HealthCheck.CloudAPI.Interval <= 0 {
			config.Cells.HealthCheck.CloudAPI.Interval = defaultCloudAPIHealthCheckInterval
		}
	}
}

func ConfigureCloud(configFile *ServerConfigFile, controllerID, nametag string) (cloud.CloudClient, error) {
	cloudClient, err := configureCloudProvider(configFile, controllerID, nametag)
	if err != nil {
		return nil, fmt.Errorf("Error setting up cloud client: %v", err)
	}

	connectWithPublicIPs := cloudClient.ConnectWithPublicIPs()
	if connectWithPublicIPs {
		klog.V(2).Infof("controller will connect to nodes via public IPs")
	} else {
		klog.V(2).Infof("controller will connect to nodes via private IPs")
	}
	err = cloudClient.EnsureKipSecurityGroups(
		configFile.Cells.ExtraCIDRs,
		configFile.Cells.ExtraSecurityGroups,
	)
	if err != nil {
		return nil, util.WrapError(err, "Error setting up cloud client security groups")
	}
	return cloudClient, err
}

const awsRegionFormat string = "[a-z]{2}-[a-z]+-[0-9]"

var awsRegionRegexp = regexp.MustCompile("^" + awsRegionFormat + "$")

func validateAWSConfig(cf *AWSConfig) field.ErrorList {
	allErrs := field.ErrorList{}

	fldPath := field.NewPath("cloud.aws")
	if cf.Region == blankTemplateValue {
		allErrs = append(allErrs, field.Required(fldPath.Child("region"), "aws region must be set or pulled from the environment"))
	} else if cf.Region != "" && !awsRegionRegexp.MatchString(cf.Region) {
		regexError := vutil.RegexError(awsRegionFormat, "us-east-1")
		allErrs = append(allErrs, field.Invalid(fldPath.Child("region"), cf.Region, regexError))
	}

	// validation of these parameters can be problematic credentials
	// can come from the config file or a ~HOME/.aws/ folder, or from
	// env vars or from an IAM role attached to an instance.

	if cf.AccessKeyID == blankTemplateValue {
		allErrs = append(allErrs, field.Required(fldPath.Child("accessKeyID"), "accessKeyID must be set or pulled from the environment"))
	}
	if cf.SecretAccessKey == blankTemplateValue {
		allErrs = append(allErrs, field.Required(fldPath.Child("secretAccessKey"), "secretAccessKey must be set or pulled from the environment"))
	}

	return allErrs
}

func validateAzureConfig(cf *AzureConfig) field.ErrorList {
	allErrs := field.ErrorList{}
	fldPath := field.NewPath("cloud.azure")

	// Items that must be set in provider.yaml
	if cf.SubscriptionID == blankTemplateValue || cf.SubscriptionID == "" {
		allErrs = append(allErrs, field.Required(fldPath.Child("subscriptionID"), "azure subscriptionID must be set in provider.yaml"))
	}
	if cf.Location == blankTemplateValue || cf.Location == "" {
		allErrs = append(allErrs, field.Required(fldPath.Child("location"), "azure Location must be set in provider.yaml"))
	}
	// Items that can be set in the Environment
	if cf.TenantID == blankTemplateValue {
		allErrs = append(allErrs, field.Required(fldPath.Child("tenantID"), "tenantID must be set in provider.yaml or pulled from the environment"))
	}
	if cf.ClientID == blankTemplateValue {
		allErrs = append(allErrs, field.Required(fldPath.Child("clientID"), "clientID must be set in provider.yaml or pulled from the environment"))
	}
	if cf.ClientSecret == blankTemplateValue {
		allErrs = append(allErrs, field.Required(fldPath.Child("clientSecret"), "clientSecret must be set in provider.yaml or pulled from the environment"))
	}

	return allErrs
}

func validateGCEConfig(cf *GCEConfig) field.ErrorList {
	allErrs := field.ErrorList{}
	fldPath := field.NewPath("cloud.gce")
	if cf.CredentialsFile != "" && cf.Credentials != nil {
		allErrs = append(allErrs, field.Invalid(fldPath, "Multiple credential sources specified", "please specify ONE of credentialsFile or credentials structure"))
	}
	if cf.CredentialsFile != "" {
		allErrs = append(allErrs, validation.ValidateFileExists(cf.CredentialsFile, fldPath.Child("credentialsFile"))...)
	}
	return allErrs
}

func validateServerConfigFile(cf *ServerConfigFile) field.ErrorList {
	allErrs := field.ErrorList{}

	cells := cf.Cells
	fldPath := field.NewPath("cells")

	if cells.CloudInitFile != "" {
		allErrs = append(allErrs, validation.ValidateFileExists(cells.CloudInitFile, fldPath.Child("cloudInitFile"))...)
	}

	allErrs = append(allErrs, validation.ValidateResourceParses(cells.DefaultVolumeSize, fldPath.Child("defaultVolumeSize"))...)

	if cells.DefaultInstanceType == "" {
		allErrs = append(allErrs, field.Required(fldPath.Child("defaultInstanceType"), ""))
	}

	if cells.StatusInterval < 0 {
		allErrs = append(allErrs, field.Invalid(fldPath.Child("statusInterval"), cells.StatusInterval, "cells.statusInterval must be >= 1"))
	}

	if cells.HealthCheck.Status != nil && cells.HealthCheck.CloudAPI != nil {
		allErrs = append(allErrs, field.Invalid(fldPath.Child("healthcheck"), "multiple healthchecks configured", "cannot set both status and cloudAPI healthchecks"))
	}
	if cells.HealthCheck.Status != nil && cells.HealthCheck.Status.HealthyTimeout <= 0 {
		allErrs = append(allErrs, field.Invalid(fldPath.Child("healthcheck.status.healthyTimeout"), cells.HealthCheck.Status.HealthyTimeout, "healthyTimeout must be positive"))
	}
	if cells.HealthCheck.CloudAPI != nil {
		if cells.HealthCheck.CloudAPI.HealthyTimeout <= 0 {
			allErrs = append(allErrs, field.Invalid(fldPath.Child("healthcheck.cloudAPI.healthyTimeout"), cells.HealthCheck.CloudAPI.HealthyTimeout, "healthyTimeout must be positive"))
		}
		if cells.HealthCheck.CloudAPI.Interval < 10 {
			allErrs = append(allErrs, field.Invalid(fldPath.Child("healthcheck.cloudAPI.interval"), cells.HealthCheck.CloudAPI.Interval, "cloudAPI interval must be greater than or equal to 10 (seconds)"))
		}
	}

	// Sadly we can't validate the default instance type until
	// after we initialize the instanceselector and the instance
	// selector needs the cloud config in order to be initialized
	// allErrs = append(allErrs, validation.ValidateInstanceType(nodes.DefaultInstanceType, fldPath.Child("defaultInstanceType"))...)

	snPath := fldPath.Child("standbyCells")
	for i, n := range cells.StandbyCells {
		snPath = snPath.Index(i)
		allErrs = append(allErrs, validation.ValidateNonnegativeField(int64(n.Count), snPath.Child("count"))...)
		if n.InstanceType == "" {
			allErrs = append(allErrs, field.Required(snPath.Child("instanceType"), ""))
		}
	}

	if len(cells.Nametag) > 0 {
		for _, msg := range validation.NameIsDNS952Label(cells.Nametag, false) {
			allErrs = append(allErrs, field.Invalid(fldPath.Child("nametag"), cells.Nametag, msg))
		}
	}

	return allErrs
}

func validateBootImageSpec(spec cloud.BootImageSpec, cloudClient cloud.CloudClient) error {
	img, err := cloudClient.GetImage(spec)
	if err != nil {
		return util.WrapError(err, "could not get machine image for %v", spec)
	}
	if img.ID == "" {
		return fmt.Errorf("could not find machine image for %v", spec)
	}
	return nil
}
